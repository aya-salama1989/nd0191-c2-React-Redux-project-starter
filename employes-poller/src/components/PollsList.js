import { connect } from "react-redux";
import PollItem from "./PollItem";
import { useState, useEffect } from "react";

const PollsList = (props) => {
  const [polls, setPolls] = useState(props.unAnsweredPolls);

  useEffect(() => {
    const userAddedQuestions = props.userQuestions.filter(
      (quetion) =>
        !props.unAnsweredPolls.includes(quetion) &&
        !props.answeredPolls.includes(quetion)
    );
    const defaultList = props.unAnsweredPolls.concat(userAddedQuestions);
    setPolls(defaultList);
  }, [props.userQuestions]);

  const showAnsweredPolls = () => {
    setPolls(props.answeredPolls);
  };

  const showUnAnsweredPolls = () => {
    setPolls(props.unAnsweredPolls);
  };

  return (
    <div>
      <div className="tab">
        <button
          className="tablinks"
          onClick={showAnsweredPolls}
          data-testid="answered-polls-test-id"
        >
          Answered Polls
        </button>
        <button
          className="tablinks"
          onClick={showUnAnsweredPolls}
          data-testid="unanswered-polls-test-id"
        >
          UnAnswered Polls
        </button>
      </div>

      <div className="lists">
        <ul>
          {polls.map((pollId) => (
            <li key={pollId} className="poll-item">
              <PollItem pollId={pollId} />
            </li>
          ))}
        </ul>
      </div>
    </div>
  );
};

const mapStateToProps = ({ polls, authedUser, users }) => {
  const user = users[authedUser];


  const authedUserAnsweredPolls =
    (user !== null || user !== undefined) && Object.keys(user.answers);

  const answeredPolls = Object.keys(polls)
    .filter((pollId) => authedUserAnsweredPolls.includes(pollId))
    .sort((a, b) => {
      return polls[b].timestamp - polls[a].timestamp;
    });

  const unAnsweredPolls = Object.keys(polls)
    .filter((pollId) => !authedUserAnsweredPolls.includes(pollId))
    .sort((a, b) => {
      return polls[b].timestamp - polls[a].timestamp;
    });

    const userQuestions = user.questions

  return {
    answeredPolls,
    unAnsweredPolls,
    userQuestions
  };
};

export default connect(mapStateToProps)(PollsList);
