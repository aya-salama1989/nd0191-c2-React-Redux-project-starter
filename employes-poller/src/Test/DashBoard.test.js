import PollsList from "../components/PollsList";
import { render, screen } from "@testing-library/react";
import * as React from "react";

import store from "../index";
import { Provider } from "react-redux";
import { MemoryRouter } from "react-router";

describe("DashBoard", () => {
  it("will catch a snapshot for DashBoard", () => {
    const view = render(
      <MemoryRouter>
        <Provider store={store}>
          <PollsList />
        </Provider>
      </MemoryRouter>
    );
    expect(view).toMatchSnapshot();
  });

  it("will have expected elements", () => {
    render(
      <MemoryRouter>
        <Provider store={store}>
          <PollsList />
        </Provider>
      </MemoryRouter>
    );
    const answeredTab = screen.getAllByTestId("answered-polls-test-id");
    expect(answeredTab).toBeInTheDocument();

    const unAnsweredTab = screen.getAllByTestId("unanswered-polls-test-id");
    expect(unAnsweredTab).toBeInTheDocument();
  });
});
