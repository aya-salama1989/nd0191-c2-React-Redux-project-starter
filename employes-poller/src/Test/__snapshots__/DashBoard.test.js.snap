// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DashBoard will catch a snapshot for DashBoard 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <div
          class="tab"
        >
          <button
            class="tablinks"
            data-testid="answered-polls-test-id"
          >
            Answered Polls
          </button>
          <button
            class="tablinks"
            data-testid="unanswered-polls-test-id"
          >
            UnAnswered Polls
          </button>
        </div>
        <div
          class="lists"
        >
          <ul>
            <li
              class="poll-item"
            >
              <a
                class="poll-list-item"
                href="/questions/xj352vofupe1dqz9emx13r"
              >
                <h4>
                  Would you rather deploy to production once every two weeks or deploy to production once every month?
                   
                </h4>
                <span>
                  created by: 
                  Mike Tsamis
                </span>
                <span>
                  creation Date: 
                  9:16:PM | 4/30/2017
                </span>
              </a>
            </li>
            <li
              class="poll-item"
            >
              <a
                class="poll-list-item"
                href="/questions/vthrdm985a262al8qx3do"
              >
                <h4>
                  Would you rather take a course on ReactJS or take a course on unit testing with Jest?
                   
                </h4>
                <span>
                  created by: 
                  Tyler McGinnis
                </span>
                <span>
                  creation Date: 
                  2:09:PM | 3/15/2017
                </span>
              </a>
            </li>
          </ul>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <div
        class="tab"
      >
        <button
          class="tablinks"
          data-testid="answered-polls-test-id"
        >
          Answered Polls
        </button>
        <button
          class="tablinks"
          data-testid="unanswered-polls-test-id"
        >
          UnAnswered Polls
        </button>
      </div>
      <div
        class="lists"
      >
        <ul>
          <li
            class="poll-item"
          >
            <a
              class="poll-list-item"
              href="/questions/xj352vofupe1dqz9emx13r"
            >
              <h4>
                Would you rather deploy to production once every two weeks or deploy to production once every month?
                 
              </h4>
              <span>
                created by: 
                Mike Tsamis
              </span>
              <span>
                creation Date: 
                9:16:PM | 4/30/2017
              </span>
            </a>
          </li>
          <li
            class="poll-item"
          >
            <a
              class="poll-list-item"
              href="/questions/vthrdm985a262al8qx3do"
            >
              <h4>
                Would you rather take a course on ReactJS or take a course on unit testing with Jest?
                 
              </h4>
              <span>
                created by: 
                Tyler McGinnis
              </span>
              <span>
                creation Date: 
                2:09:PM | 3/15/2017
              </span>
            </a>
          </li>
        </ul>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
